page_iterator.go

```diff
@@ -0,0 +1,203 @@
+	"errors"
+	"net/url"
+	"reflect"
+	"unsafe"
+
+	abstractions "github.com/microsoft/kiota/abstractions/go"
+	"github.com/microsoft/kiota/abstractions/go/serialization"
+)
+
+type Page interface {
+	getValue() []interface{}
+	getNextLink() *string
+}
+
+// PageIterator represents an iterator object that can be used to get subsequent pages of a collection.
+type PageIterator struct {
+	currentPage     Page
+	reqAdapter      GraphRequestAdapterBase
+	pauseIndex      int
+	constructorFunc ParsableConstructor
+	headers         map[string]string
+}
+
+type ParsableConstructor func() serialization.Parsable
+
+type PageResult struct {
+	nextLink *string
+	value    []interface{}
+}
+
+func (p *PageResult) getValue() []interface{} {
+	if p == nil {
+		return nil
+	}
+
+	return p.value
+}
+
+func (p *PageResult) getNextLink() *string {
+	if p == nil {
+		return nil
+	}
+
+	return p.nextLink
+}
+
+// NewpageIterator creates an iterator instance
+//
+// It has three parameters. res is the graph response from the initial request and represents the first page.
+// reqAdapter is used for getting the next page and constructorFunc is used for serializing next page's response to the specified type.
+func NewPageIterator(res interface{}, reqAdapter GraphRequestAdapterBase, constructorFunc ParsableConstructor) (*PageIterator, error) {
+	page, err := convertToPage(res)
+	if err != nil {
+		return nil, err
+	}
+
+	return &PageIterator{
+		currentPage:     page,
+		reqAdapter:      reqAdapter,
+		pauseIndex:      0,
+		constructorFunc: constructorFunc,
+		headers:         map[string]string{},
+	}, nil
+}
+
+// Iterate traverses all pages and enumerates all items in the current page and returns an error if something goes wrong.
+//
+// Iterate receives a callback function which is called with each item in the current page as an argument. The callback function
+// returns a boolean. To traverse and enumerate all pages always return true and to pause traversal and enumeration
+// return false from the callback.
+//
+// Example
+//      pageIterator, err := NewPageIterator(resp, reqAdapter, parsableCons)
+//      callbackFunc := func (pageItem interface{}) bool {
+//          fmt.Println(pageitem.GetDisplayName())
+//          return true
+//      }
+//      err := pageIterator.Iterate(callbackFunc)
+func (pI *PageIterator) Iterate(callback func(pageItem interface{}) bool) error {
+	for pI.currentPage != nil {
+		keepIterating := pI.enumerate(callback)
+
+		if !keepIterating {
+			// Callback returned false, stop iterating through pages.
+			return nil
+		}
+
+		err := pI.next()
+		if err != nil {
+			return err
+		}
+		pI.pauseIndex = 0 // when moving to the next page reset pauseIndex
+	}
+
+	return nil
+}
+
+// SetHeaders provides headers for requests made to get subsequent pages
+//
+// Headers in the initial request -- request to get the first page -- are not included in subsequent page requests.
+func (pI *PageIterator) SetHeaders(headers map[string]string) {
```

Comment:
for completeness we should probably also have a set options method

Reply:
The passed instance of request adapter has options set, why should we have set options on the iterator as well?

Reply:
to be more specific I'm referring to request options that get passed to handlers. You can override the default options for each request, look at the O field on the options objects (yep using the options word twice was probably not such a great idea after all)